library/Freckle/App/Kafka.hs

```diff
@@ -0,0 +1,132 @@
+import Data.ByteString (toStrict)
+import qualified Data.List.NonEmpty as NE
+import Data.Pool (Pool)
+import qualified Data.Pool as Pool
+import qualified Data.Text as T
+import qualified Freckle.App.Env as Env
+import qualified Kafka.Producer as Kafka
+import UnliftIO.Async (async)
+import UnliftIO.Exception (throwString)
+import Yesod.Core.Lens
+import Yesod.Core.Types (HandlerData)
+
+newtype BrokerAddress = BrokerAddress { getBrokerAddress :: Maybe (NonEmpty Kafka.BrokerAddress)}
+
+class HasKafkaBrokerAddress env where
+  kafkaBrokerAddressL :: Lens' env BrokerAddress
+
+instance HasKafkaBrokerAddress site => HasKafkaBrokerAddress (HandlerData child site) where
+  kafkaBrokerAddressL = envL . siteL . kafkaBrokerAddressL
+
+envKafkaBrokerAddress
+  :: Env.Parser Env.Error BrokerAddress
+envKafkaBrokerAddress = BrokerAddress <$> addressOrNothing
+
+ where
+  addressOrNothing = (parseKafkaBrokerAddress <$> parseKey) <|> pure Nothing
+  parseKey = Env.var Env.nonempty "KAFKA_BROKER_ADDRESS" mempty
+  parseKafkaBrokerAddress = NE.nonEmpty . fmap Kafka.BrokerAddress . T.splitOn ","
+
+data KafkaProducer
+  = NullKafkaProducer
+  | KafkaProducerPool (Pool Kafka.KafkaProducer)
+
+class HasKafkaProducer env where
+  kafkaProducerL :: Lens' env KafkaProducer
+
+instance HasKafkaProducer site => HasKafkaProducer (HandlerData child site) where
+  kafkaProducerL = envL . siteL . kafkaProducerL
+
+kafkaProducer
+  :: ( MonadReader env m
+     , HasKafkaBrokerAddress env
+     , MonadUnliftIO m
+     )
+  => Int
+  -- ^ The number of stripes (distinct sub-pools) to maintain.
+  -- The smallest acceptable value is 1.
+  -> NominalDiffTime
+  -- ^ Amount of time for which an unused resource is kept open.
+  -- The smallest acceptable value is 0.5 seconds.
+  --
+  -- The elapsed time before destroying a resource may be a little
+  -- longer than requested, as the reaper thread wakes at 1-second
+  -- intervals.
+  -> Int
+  -- ^ Maximum number of resources to keep open per stripe.  The
+  -- smallest acceptable value is 1.
+  --
+  -- Requests for resources will block if this limit is reached on a
+  -- single stripe, even if other stripes have idle resources
+  -- available.
+  -> m KafkaProducer
+kafkaProducer numStripes idleTime maxResources =
+  view kafkaBrokerAddressL >>= \case
+    BrokerAddress Nothing -> pure NullKafkaProducer
+    BrokerAddress (Just addresses) -> do
+      let
+        pool =
+          Pool.createPool mkProducer Kafka.closeProducer numStripes idleTime maxResources
+        mkProducer =
+          either throw pure =<< Kafka.newProducer (Kafka.brokersList $ toList addresses)
+        throw err = throwString $ "Failed to open kafka producer: " <> show err
+      KafkaProducerPool
+        <$> liftIO pool
+
+produceAsync
+  :: ( ToJSON value
+     , ToJSON key
+     , MonadLogger m
+     , MonadReader env m
+     , HasKafkaProducer env
+     , MonadUnliftIO m
+     )
+  => Kafka.TopicName
+  -> NonEmpty value
+  -> (value -> key)
```

Comment:
This will be an interesting design point vs a `HasKafkaPartitionKey` class on `value`. On the one hand, the class would make it easy to decide partitioning in one place, where the event data type is created. Just passing a function on every send is pretty easy.

You could also offer all the options though...

```hs
produce :: HasKafkaPartitionKey value => value -> 
produce = produceKeyedBy kafkaPartitionKey

produceKeyedBy :: (value -> key) -> value ->
produceKeyedBy f v = produceKeyed (f v) v

produceKeyed :: key -> value ->
produceKeyed k v = do
```

Reply:
One option is to name what you have `produceKeyedBy`, so as not to squat on `produce(Async)` for something class-based later.

Reply:
The class based approach would force the `key` to always depend on the `value` which is probably not always the case. I'm passing `const studentId` in the current usage for example.

I went with `produceKeyedBy` et al. to keep things flexible for the future.

Reply:
> I'm passing `const studentId` in the current usage for example

Sounds like you _actually_ want `produceKeyed studentId` to be the first function we have then.

But I'm very surprised `AnswerEvent` lacks a `studentId` property. Is that right? We have no need for consumers to know the student that generated an answer? I would've really expected it to, and for "always key answer events by student" to be _exactly_ the sort of thing you'd ideally define in a single place, say as an instance on `AnswerEvent`.

Even still, you could always...

```hs
data ConstantKey k a = ConstantKey k a

instance ToJSON a => ToJSON (ConstantKey k a) where
  toJSON (ConstantKey _ a) = toJSON a

instance ToKafkaPartitionKey (ConstantKey k a) where
  toKafkaPartitionKey (ConstantKey k _) = toKafkaPartitionKey a

produce $ ConstantKey studentId <$> answerEvents
```

You're totally right, of course. Not having the `Keyed` or `KeyedBy` functions available too would be a huge limitation for the library. The class-based functions would be the additional convenience for cases where that made sense.